### DISCUSSION TITLE
Data section improvements

### DISCUSSION THREAD

**Discussion by danielerpl** [DIY-Flow-Bench/DIY-Flow-Bench#241] [Code Development]
@DeeEmm  I am making some improvement to the data section, like mentioned to you at the beginning of this tour with your amazing product. What I am doing now is, make the possibility to upload 2 lift files to have a comparison chart. There is also the possibility to download the final compared chart. 
I added this section below the original chart, to have everything in one place.
The values and the legend of the chart it's dynamic, based on the input files. I would also make this feature to the original chart because I think we can have different cfm point values. 
This is a screenshot of what I am doing, I will push the code soon when I completed the entire functionality. 
Let me know what do you think about, we can make some adjustment and improvements 

<img width="1026" alt="Screenshot 2024-11-19 at 10 32 44" src="https://github.com/user-attachments/assets/20798080-850c-44d8-b852-04fabd445dbb">

**Comment by DeeEmm** [comment]
That's awesome Daniele. Nice work. 👍 

The ability to compare two traces is a very useful tool.

I am not sure if you have been following the recent changes, there is now a DEV branch that contains the latest code. All changes should be made using the DEV branch as the base. The documentation may be a little out of date on this particular workflow.

The plan moving forwards is that the master branch will be left at the same version as the last release. Which will allow people to download the code either from the release page or by forking / downloading the master branch.

All development is now based on the DEV branch. Forks can be created on issues, but the DEV branch needs to be manually set as the source for the fork. 

For a new release: once all PRs are in and the DEV branch has been tested it can be merged with the Master to update it and create a new release.

**Reply by danielerpl** [reply]
Yes it is very helpful to make a properly evidence of the changes. 
No problem for the repo, I will fork the dev repo and I will make the pull request there 😉 The functionality it's 100% inside the index.html this will make much easier the implementation to the DEV repo.
I will keep you updated about this develop. If you have any suggestion I am here to make it!!

**Reply by DeeEmm** [reply]
> The functionality it's 100% inside the index.html this will make much easier the implementation to the DEV repo.

Yes, that does make the PR very easy 😄 

I am updating the projects Contributions file to capture the changes and also looking in to better ways to manage changes. Also looking in to some YAML workflow integrations to automate some of the process, which is a new thing for me.

With regards to the Data Graph. There are lots of future plans. Amongst the best suggestions is the ability to display extended data below the graph in the form of a Data Grid. This way for each data point on the graph there will be a matching column below that shows the Flow Value, Depression, Adjusted Flow Value, etc. 

Of course half of this change would be within the main code, but I am happy to collaborate on that once we get the next release out of the way.

**Comment by danielerpl** [comment]
@DeeEmm I would like to move ahead and make the implementation for the automatic pressure adjustment  during the test ( 28 h2o) in a previous discussion you mentioned, the code it's already prepared for this. About the hardware, do you know the best component to use for this case? I would start to prepare my bench for this new implementation, and then move to the code improvements. What do you think about it ?

**Reply by DeeEmm** [reply]
At the moment my focus is on getting the V2 version released. Features such as automatic depression regulation are part of V3 development which is on hold until the V2 release is completed.

Otherwise it gets too complicated to manage the code changes as development of such a feature is undertaken on outdated code. It also ends up delaying the release of current stable code.

Once the V2 release is done, we will have a vote on which V3 features get priority.

I will publish the RC9 release over the next day or so. I only have the demo on the website to update. Hopefully the official V2 release is not too far after that.

**Reply by danielerpl** [reply]
@DeeEmm I am looking to add the automatic pressure adjustment on my current board, the version V2. I am looking to use the  Adafruit MCP4725 module. Do we have available ports to connect it to the board ? I am pretty sure we don't have available I2C. Can you confirm ?

**Reply by DeeEmm** [reply]
Hi Daniel, Sorry for my late reply. I've been tied up with work stuff.

What is the intent with using the MCP4725  ADC for depression control?

Existing intent is to use the PWM outputs to control depression, either as a direct input to a VFD, or via bridge circuit. (X2_VAC_BLEED_VALVE_REF) 

OR

Use the existing stepper control to directly control a bleed valve. (X2_FC_STEP_)

OR

A combination of both.

The existing pRef sensor provides the depression value, so in the code we will implement a PID control to modulate the VFD / Bleed valve based on the pRef value

We also have motor bank control (X2_MOTOR_BANK_1/2/3) to control staged vac motor banks that can be utilised as part of the overall control strategy

**Reply by DeeEmm** [reply]
<img width="1238" height="281" alt="Screenshot 2025-10-19 at 10 48 29 am" src="https://github.com/user-attachments/assets/72424ad4-1dec-4f4f-90a8-8ca6f74b1dfe" />

**Reply by danielerpl** [reply]
Don't worry for the late reply! ;) 

Currently I am using the analog potentiometer to manipulate the speed of the motors. If I have one or six motors turned on, I am able to speed up or down with this analog potentiometer. 

The idea is to use an automatic potentiometer to make changes for the motors, based on the the value configured. For example I will make set point for 28 cfm and the board automatically will change the speed of the motors. 

My idea was to use this MCP4725 to give the 0-10v to the motors controller like the current potentiometer. 

Do you already implemented the functionality in the current V2 board ? 
Please can you give me more information or share the link where It's documented ? 
Many thanks

**Reply by DeeEmm** [reply]
No, the motor control is on the roadmap, but development has not started yet. I've been tied up with life stuff and so stepped away from development of the project for a while. With the basic code and hardware functional it seemed like a good place to leave things for a while.

The I/O above was determined back when the PCB was designed. That way the hardware / PCB would not need to be changed as the various I/O are already in place just waiting for the code to drive it. It's just forward planning of what potential use cases there may be. So things like stepper controlled bleed valves and automatic valve adjustment were identified early on. So the 'intent' is as described, but the actual code required to drive that functionality has not been developed at this point in time.

Of course, as none of those items are part of the core operation, the I/O can also easily be reassigned for other uses, should the end user want.

IIRC we were going to have a vote to see what feature people wanted to see next,  but I think that closed loop auto modulation of vac control sources is the most likely. PID is a standard control philosophy. The PWM output is a 0-3.3v output, so that would need to be 'stepped up' to match whatever was being controlled. i.e. 0-10v, 4-20mA, 0-5v, etc. The actual conversion would therefore be handled by intermediate circuitry to match whatever it was the end user is using. Be that a VFD or 240v PWM or whatever. 

Essentially the PWM output control would be the same in all cases, it's modulated based on the pRef signal. So the target is set on the GUI (i.e 28") and when the motors are started, the PWM output is modulated to ensure that 28" is maintained (using PID control). If it cannot achieve 28", then it will simply max out at 100% duty. Then if multiple banks of motors are enabled it can bring in another bank, et cetera.

It is of course possible to use the MCP4725. The I2C bus is a non terminated bus, so you can simply tap into the existing SDA / SCL lines and as long as the code is present to communicate with it, it will work. But with that said, as the PWM outputs of the ESP32 are also 12 bit, and the MCP4725 is only a 0-5v device, there does not seem to be much of an advantage to using it. It has a shared supply, so can only put out a max of 5v if connected to the 5v side of the I2C bus. So to be able to control a 0-10v device you still need to use some kind of intermediate control circuit

TBH. There are already widely available tunable PWM driven BOBs that will take a 0-3.3v reference and output a variety of voltages. It would be trivial to integrate one of these by tapping in to the 12v supply and driving from the existing PWM output to give you a 0-10v output. They are commonly used on things like BDM motor and laser control and are pretty cheap...

Here's an example...

<img width="1684" height="744" alt="Screenshot 2025-10-20 at 11 00 11 am" src="https://github.com/user-attachments/assets/26c78213-b9a0-4a5a-b884-c15ba1e0fd46" />

Of course this is just the first result from a random search, so not specifically promoting this device, but from a project perspective, these are cheap and easy to obtain and the end user does not need to dive into circuit level design. Of course, it is also possible to use the prototype area of the PCB to do this if the end user should want.

So, some food for thought.

**Comment by danielerpl** [comment]
Hi Mick

I very  appreciated all the info you shared with me about this new
improvement for the project. I will continue the investigation about what I
am receiving from the DIY project and what I have in my bech. I will keep
you posted about the progress on it.

Daniele

**Comment by danielerpl** [comment]
Hello @DeeEmm after I while, I completed the board for the pressure tracker to be connected on your board. The board I made should control all my 6 motor based on the h2o reference value selected from the dashboard console. 
I will share you some info once it’s connected and the job completed! I will keep you updated. At the moment the board is on the way of the production!!

**Reply by DeeEmm** [reply]
Cool, that sounds awesome, and timely too.

Just getting back to looking at this project again, so could be an opportunity to roll some of this into the main repo.

Keep us posted.
